yeat/workflows/workflows.py

```diff
@@ -0,0 +1,105 @@
+# -------------------------------------------------------------------------------------------------
+# Copyright (c) 2023, DHS. This file is part of YEAT: http://github.com/bioforensics/yeat
+#
+# This software was prepared for the Department of Homeland Security (DHS) by the Battelle National
+# Biodefense Institute, LLC (BNBI) as part of contract HSHQDC-15-C-00064 to manage and operate the
+# National Biodefense Analysis and Countermeasures Center (NBACC), a Federally Funded Research and
+# Development Center.
+# -------------------------------------------------------------------------------------------------
+
+from pathlib import Path
+from pkg_resources import resource_filename
+from snakemake import snakemake
+
+
+def run_paired(
+    fastq1,
+    fastq2,
+    assembly_configs,
+    outdir=".",
+    cores=1,
+    sample="sample",
+    dryrun="dry",
+    downsample=0,
+    coverage=150,
+    seed=None,
+    genomesize=0,
+):
+    snakefile = resource_filename("yeat", "workflows/snakefiles/Paired")
+    r1 = Path(fastq1).resolve()
+    if not r1.is_file():
+        raise FileNotFoundError(f"No such file: '{r1}'")
+    r2 = Path(fastq2).resolve()
+    if not r2.is_file():
+        raise FileNotFoundError(f"No such file: '{r2}'")
+    assemblers = [config.algorithm for config in assembly_configs]
+    extra_args = {config.algorithm: config.extra_args for config in assembly_configs}
+    config = dict(
+        read1=r1,
+        read2=r2,
+        assemblers=assemblers,
+        extra_args=extra_args,
+        sample=sample,
+        threads=cores,
+        downsample=downsample,
+        coverage=coverage,
+        seed=seed,
+        genomesize=genomesize,
+    )
+    success = snakemake(
+        snakefile, config=config, cores=cores, dryrun=dryrun, printshellcmds=True, workdir=outdir
+    )
+    if not success:
+        raise RuntimeError("Snakemake Failed")
+
+
+def check_canu_required_params(extra_args, cores):
+    if "genomeSize=" not in extra_args:
+        raise ValueError("Missing required input argument from config: 'genomeSize'")
+    if cores < 4:
+        raise ValueError(
+            "Canu requires at least 4 avaliable cores; increase `--threads` to 4 or more"
+        )
```

Comment:
In order to run Canu, users must input the `genomeSize` of the sample.

| CLI parameter  | Input | Does YEAT provide automatically?
| ------------- | ------------- | ------------- |
| genomeSize=  | An estimate of the size of the genome. Common suffices are allowed, for example, 3.7m or 2.8g.  | No, users will need to provide this data in the `extra_params` "Canu" section in the config |

Users cannot use YEAT's `-g` or `--genome-size` because YEAT cannot handle shortcut suffices of genomes sizes--only the int allowed.

Furthermore, Canu requires **at least 4** available cores to run. If the computer, job, etc. does not have enough, Canu will fail to run.

Added a check statement making sure that the user puts at least 4 cores when filling out YEAT's CLI with `-t` or `--threads`.

Reply:
Out of curiosity: does the Mash-based genome size estimation not work for hifi data?

Reply:
I haven't tried yet! I completely forgot about that. Sean and I discussed about it and we both think it should work! I'll do some testing.

Reply:
UPDATE: Did some testing with Mash's genome size estimation. Calculated genome sizes way over the actual genome size. Tried to adjust `mash sketch` parameters but was unable to get to the actual size.

The dataset that I used was PacBio's HiFi reads for [Ecoli K12](https://trace.ncbi.nlm.nih.gov/Traces/index.html?view=run_browser&acc=SRR10971019&display=metadata). Ecoli K12 has a genome size of ~4.6 mb (see paper [here](https://pubmed.ncbi.nlm.nih.gov/9278503/)). It is a very well studied bacteria. 

Theoretically, we should be able to calculate the genome size with PacBio reads and with high confidence. To calculate the genome size, we take the total number of k-mers / coverage. PacBio reads have a 99.9999% error-free rate. Because of this, we would expect a high accuracy in the total number of k-mers.

When running `mash sketch`, I got back a variety of over-estimated genome sizes. I adjusted the parameters to get to the expected size but was unable to.

Here are the results of my mash sketch commands:

| command  | parameters | estimated genome size | coverage |
| ------------- | ------------- | ------------- | ------------- |
| mash sketch [fastq]  | -r  | 33.6 mb | 40x |
| mash sketch [fastq]  | -k 32 -r  | 51.3 mb | 24x |
| mash sketch [fastq]  | -k 32 -m  3 -r  | 11 mb | 117x |
| mash sketch [fastq]  | -k 21 -m 3 -r  | 8.7 mb | 144x |

-r = read set
-k = k-mer size
-m = Minimum copies of each k-mer required to pass noise filter for reads.

Not sure why I am unable to get close to the expected genome size. Consulted with @lovettse, but was unable to figure it out.

---

To do:
1. try another dataset of pacbio hifi read data with known genome size to check the genome size estimation. 
2. test Illumina short read data and check if I can get the expected genome size estimation (did not test this in the past)
3. how much of an impact will a genome size with x2 or x3 the actual size will affect Canu's assembly? (Canu requires genome size.)
4. sourmash - a variant of MASH?

@lovettse mentioned something interest that I didn't think about before. Most bacteria genomes is around 7 kb. Maybe if the input data is bacteria we can set that value as default? (Very duck-tape answer.)

Reply:
Created issue thread #42 . I will revisit this problem in another PR. For this PR, let's keep it the way it is. I'll delve into estimating genome size problem in the future and integrate it in the workflow.